app/models/template_input.rb

```diff
@@ -29,4 +29,120 @@ def variable_template_input?
   def puppet_parameter_template_input?
     self.input_type == 'puppet_parameter'
   end
+
+  def preview(renderer)
+    get_resolver(renderer).preview
+  end
+
+  def value(renderer)
+    get_resolver(renderer).value
+  end
+
+  private
+
+  def get_resolver(renderer)
+    resolver_class = case input_type
+      when 'user'
+        UserInputResolver
+      when 'fact'
+        FactInputResolver
+      when 'variable'
+        VariableInputResolver
+      when 'puppet_parameter'
+        PuppetParameterInputResolver
+      else
+        raise "unknown template input type #{input_type.inspect}"
+      end
+    resolver_class.new(self, renderer)
+  end
+
+  class InputResolver
+    def initialize(input, renderer)
+      @input = input
+      @renderer = renderer
+    end
+
+    def preview
+      self.ready? ? self.resolved_value : self.preview_value
+    end
+
+    def value
+      self.ready? ? self.resolved_value : raise(ValueNotReady, "Input '#{@input.name}' is not ready for rendering")
+    end
+
+    def preview_value
+      "$#{@input.input_type.upcase}_INPUT[#{@input.name}]"
+    end
+
+    # should be defined in descendants
+    def ready?
+      raise NotImplementedError
+    end
+
+    # should be defined in descendants
+    def resolved_value
+      raise NotImplementedError
+    end
+  end
+
+  class UserInputResolver < InputResolver
```

Comment:
I'd like to keep it this way, `InputResolver` is internal object that should not be used out of `TemplateInput` scope. Maybe extracting `InputResolver` with all it's children to extra file could be a compromise but I don't see a benefit right now. (note that in Foreman small classes like exceptions, safe mode jails, validators are also in one file)